# <span style="color:teal;">CIS 211 Project 7:  SQLite</span>

##### Due 11:00 P.M. Tuesday Mar 7

This project will give you some experience writing database queries in SQL and connecting to a SQLite database from Python.

### <span style="color:teal">Setup</span>

Download the Sakila database (the file named `sakila211.db`) and move it to the same folder as this IPython notebook.

### <span style="color:teal">Executing Queries in an IPython Notebook</span>

To solve the problems in the first two sections of this project you need to write a SQL query based on the tables in the Sakila211 database.  

To execute a query from an IPython Notebook write the query in quotes as part of a shell command:
```
! sqlite3 sakila211.db "..."
```

For example, to answer the question "What are the names and ratings of the films" your code cell would have
```
! sqlite3 sakila211.db "select title, rating from film"
```

Some of the queries will produce a lot of output.  You can add a `LIMIT` clause to any query if you want to reduce the size of the output.

In [1]:
! sqlite3 sakila211.db "select title, rating from film limit 5"

title             rating    
----------------  ----------
Academy Dinosaur  PG        
Ace Goldfinger    G         
Adaptation Holes  NC-17     
Affair Prejudice  G         
African Egg       G         


In [2]:
! sqlite3 sakila211.db .tables

actor          country        film_category  payment      
address        customer       film_text      rental       
category       film           inventory      staff        
city           film_actor     language       store        


In [3]:
! sqlite3 sakila211.db .schema country;

CREATE TABLE store (
  store_id INT NOT NULL,
  manager_staff_id SMALLINT NOT NULL,
  address_id INT NOT NULL,
  PRIMARY KEY  (store_id)
);
CREATE TABLE staff (
  staff_id SMALLINT NOT NULL,
  first_name VARCHAR(45) NOT NULL,
  last_name VARCHAR(45) NOT NULL,
  address_id INT NOT NULL,
  picture BLOB DEFAULT NULL,
  email VARCHAR(50) DEFAULT NULL,
  store_id INT NOT NULL,
  active SMALLINT DEFAULT 1 NOT NULL,
  username VARCHAR(16) NOT NULL,
  password VARCHAR(40) DEFAULT NULL,
  PRIMARY KEY  (staff_id)
);
CREATE TABLE rental (
  rental_id INT NOT NULL,
  rental_date TIMESTAMP NOT NULL,
  inventory_id INT  NOT NULL,
  customer_id INT  NOT NULL,
  return_date TIMESTAMP DEFAULT NULL,
  staff_id SMALLINT  NOT NULL,
  PRIMARY KEY (rental_id)
);
CREATE TABLE payment (
  payment_id int NOT NULL,
  customer_id INT  NOT NULL,
  staff_id SMALLINT NOT NULL,
  rental_id INT DEFAULT NULL,
  amount DECIMAL(5,2) NOT NULL,
  payment_date TIMESTAMP NOT NULL,
  PRIMARY KEY  (payment_id)
);
CREATE TABLE

Error: near "country": syntax error


**Note:** &nbsp; Your grade will not depend on the output format.  If want to have a nicer looking output that includes column headers create a file named `.sqliterc` in your home folder and add these two lines:
```
.header on
.mode column
```

## <span style="color:teal">Part 1:  &nbsp; Queries</span>

The following questions can all be answered using records from a single table.

(1) What are the first and last names of all the customers?

In [4]:
! sqlite3 sakila211.db "SELECT * FROM store, staff LIMIT 5"

store_id    manager_staff_id  address_id  staff_id    first_name  last_name   address_id  picture     email                         store_id    active      username    password                                
----------  ----------------  ----------  ----------  ----------  ----------  ----------  ----------  ----------------------------  ----------  ----------  ----------  ----------------------------------------
1           1                 1           1           Mike        Hillyer     3                       Mike.Hillyer@sakilastaff.com  1           1           Mike        8cb2237d0679ca88db6464eac60da96345513964
1           1                 1           2           Jon         Stephens    4                       Jon.Stephens@sakilastaff.com  2           1           Jon         8cb2237d0679ca88db6464eac60da96345513964
2           2                 2           1           Mike        Hillyer     3                       Mike.Hillyer@sakilastaff.com  1           1           Mike    

In [5]:
! sqlite3 sakila211.db "SELECT first_name, last_name FROM customer"

first_name  last_name 
----------  ----------
Mary        Smith     
Patricia    Johnson   
Linda       Williams  
Barbara     Jones     
Elizabeth   Brown     
Jennifer    Davis     
Maria       Miller    
Susan       Wilson    
Margaret    Moore     
Dorothy     Taylor    
Lisa        Anderson  
Nancy       Thomas    
Karen       Jackson   
Betty       White     
Helen       Harris    
Sandra      Martin    
Donna       Thompson  
Carol       Garcia    
Ruth        Martinez  
Sharon      Robinson  
Michelle    Clark     
Laura       Rodriguez 
Sarah       Lewis     
Kimberly    Lee       
Deborah     Walker    
Jessica     Hall      
Shirley     Allen     
Cynthia     Young     
Angela      Hernandez 
Melissa     King      
Brenda      Wright    
Amy         Lopez     
Anna        Hill      
Rebecca     Scott     
Virginia    Green     
Kathleen    Adams     
Pamela      Baker     
Martha      Gonzalez  
Debra       Nelson    
Amanda      Carter    
Stephanie   Mitchell  
Carolyn    

(2) What are the e-mail addresses of customers whose first name is “Willard”?

In [6]:
! sqlite3 sakila211.db "SELECT email FROM customer WHERE first_name = 'Willard'"

email                             
----------------------------------
willard.lumpkin@sakilacustomer.org


(3) How many customers are there?

In [7]:
! sqlite3 sakila211.db "SELECT count(*) FROM customer"

count(*)  
----------
599       


(4) How many customers shop at store number 1?

In [8]:
! sqlite3 sakila211.db "SELECT count(*) FROM customer where store_id = 1"

count(*)  
----------
326       


(5) How much does it cost to rent the film named “Virtual Spoilers”?

In [9]:
! sqlite3 sakila211.db "SELECT rental_rate FROM film WHERE title = 'Virtual Spoilers'"

rental_rate
-----------
4.99       


(6) Do any films have the word “Princess” in the title?

In [10]:
! sqlite3 sakila211.db "SELECT title FROM film WHERE title LIKE '%Princess%'"

title         
--------------
Image Princess
Princess Giant


(7) What are the titles of the films that are longer than 180 minutes?

In [11]:
! sqlite3 sakila211.db "SELECT title FROM film WHERE length > 180"

title           
----------------
Analyze Hoosiers
Baked Cleopatra 
Catch Amistad   
Chicago North   
Conspiracy Spiri
Control Anthem  
Crystal Breaking
Darn Forrester  
Frontier Cabin  
Gangs Pride     
Haunting Pianist
Home Pity       
Hotel Happiness 
Intrigue Worst  
Jacket Frisco   
King Evolution  
Lawless Vision  
Love Suicides   
Monsoon Cause   
Moonwalker Fool 
Muscle Bright   
Pond Seattle    
Records Zorro   
Reds Pocus      
Runaway Tenenbau
Saturn Name     
Scalawag Duck   
Searchers Wait  
Smoochy Control 
Soldiers Evoluti
Sons Interview  
Sorority Queen  
Star Operation  
Sweet Brotherhoo
Theory Mermaid  
Wife Turn       
Wild Apollo     
Worst Banger    
Young Language  


(8) How many films have a “G” rating and are less than 60 minutes long?

In [12]:
! sqlite3 sakila211.db "SELECT count(*) FROM film WHERE rating = 'G' AND length < 60"

count(*)  
----------
21        


(9) What is the maximum replacement cost for any film?

In [13]:
! sqlite3 sakila211.db "SELECT max(replacement_cost) FROM film"

max(replacement_cost)
---------------------
29.99                


(10) Print a table that lists the different types of ratings and the number of films that have that rating.

In [14]:
! sqlite3 sakila211.db "SELECT rating, count(*) FROM film GROUP BY rating"

rating      count(*)  
----------  ----------
G           178       
NC-17       210       
PG          194       
PG-13       223       
R           195       


## <span style="color:teal">Part 2: &nbsp; Joins</span>

The following queries all require a join of two or more tables. As a hint for how to create the query the table names are shown next to each a question.

(11) What are the film IDs of movies starring Jude Cruise?  &nbsp; &nbsp; [`actor`, `film_actor`]

In [15]:
! sqlite3 sakila211.db "SELECT film_id FROM film_actor JOIN actor USING (actor_id) WHERE first_name = 'Jude' AND last_name = 'Cruise'"

film_id   
----------
16        
34        
101       
114       
122       
134       
144       
153       
192       
213       
258       
267       
317       
340       
393       
437       
447       
502       
592       
605       
637       
685       
707       
714       
717       
737       
767       
852       
891       
918       


(12) What language was “Ace Goldfinger” filmed in?  &nbsp; &nbsp; [`film`, `language`]

In [16]:
! sqlite3 sakila211.db "SELECT name as Language FROM language JOIN film USING (language_id) WHERE title = 'Ace Goldfinger'"

Language  
----------
Japanese  


(13) How many actors starred in the film named “Scalawag Duck”? &nbsp; &nbsp; [`film`, `film_actor`]

In [17]:
! sqlite3 sakila211.db "SELECT count(*) as Actor_Count FROM film_actor JOIN film USING (film_id) WHERE title = 'Scalawag Duck'"

Actor_Count
-----------
5          


(14) Produce a table that shows how many films each customer has rented.  &nbsp; &nbsp; [`customer`, `rental`]

In [18]:
! sqlite3 sakila211.db "SELECT customer_id, count(customer_id) AS films_rented FROM rental JOIN customer USING (customer_id) GROUP BY customer_id"

customer_id  films_rented
-----------  ------------
1            32          
2            27          
3            26          
4            22          
5            37          
6            28          
7            33          
8            24          
9            22          
10           25          
11           23          
12           28          
13           27          
14           27          
15           30          
16           28          
17           21          
18           22          
19           24          
20           30          
21           34          
22           21          
23           29          
24           25          
25           29          
26           34          
27           31          
28           31          
29           35          
30           34          
31           26          
32           29          
33           20          
34           24          
35           32          
36           27          
37          

(15) How many films did the customer named Smith rent?  &nbsp; &nbsp; [`customer`, `rental`]

In [19]:
! sqlite3 sakila211.db "SELECT last_name, count(customer_id) AS films_rented FROM rental JOIN customer USING (customer_id) WHERE last_name = 'Smith'"

last_name   films_rented
----------  ------------
Smith       32          


(16) Which actors starred in “Splash Gump”?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [20]:
! sqlite3 sakila211.db "SELECT first_name, last_name FROM actor JOIN film_actor USING (actor_id) JOIN film USING (film_id) WHERE title = 'Splash Gump'"

first_name  last_name 
----------  ----------
Penelope    Guiness   
Jennifer    Davis     
Uma         Wood      
Cameron     Streep    
Christian   Neeson    
Ray         Johansson 
Gene        Willis    
Morgan      Williams  
Humphrey    Willis    
Al          Garland   


(17) Which films has Penelope Guiness starred in?  &nbsp; &nbsp; [`film`, `film_actor`, `actor`]

In [21]:
! sqlite3 sakila211.db "SELECT first_name, last_name, count(actor_id) AS '# of films' FROM film_actor JOIN actor USING (actor_id) JOIN film USING (film_id) WHERE first_name = 'Penelope' AND last_name = 'Guiness'"

first_name  last_name   # of films
----------  ----------  ----------
Penelope    Guiness     19        


(18) What are the names of the films rented by the customer named Knight?  &nbsp; &nbsp; [`customer`, `rental`, `inventory`, `film`]

In [22]:
! sqlite3 sakila211.db "SELECT title FROM film JOIN inventory USING (film_id) JOIN rental USING (inventory_id) JOIN customer USING (customer_id) WHERE last_name = 'Knight'"

title        
-------------
Jericho Mulan
Frost Head   
Clerks Angels
Wife Turn    
Expecations N
Lawless Visio
Soup Wisdom  
Heartbreakers
Bikini Borrow
Attacks Hate 
Cupboard Sinn
Princess Gian
Cyclone Famil
Platoon Insti
Knock Warlock
Excitement Ev
Right Cranes 
Fatal Haunted
Falcon Volume
Tomorrow Hust
Half Outfield
Horror Reign 
Borrowers Bed


## <span style="color:teal">Part 3: &nbsp; Monthly Report</span>

For this part of the project you will write a report generator that will print a summary of monthly activity for a customer.  The report should show a list of all the films rented during the month along with a total amount due.
This example shows the activity for the customer named Leo Ebert for the month of June, 2005.

<pre>
--- Sakila DVD Rentals ---

Monthly Report for Leo Ebert 

Lonely Elephant                 06/16/2005    $2.99
                   ** late fee  06/24/2005    $2.99
Enemy Odds                      06/18/2005    $4.99
                   ** late fee  06/27/2005    $4.99
Song Hedwig                     06/20/2005    $0.99
                   ** late fee  06/26/2005    $0.99
Wash Heavenly                   06/21/2005    $4.99
Hurricane Affair                06/21/2005    $2.99

Monthly Total:  $25.92
</pre>

To compute the total, the report generator needs to add up all the rental fees, and for any film returned late, add an additional rental fee as the late charge. In the example, Leo rented *Lonely Elephant* on June 16 but didn’t return it until June 24. It was a 3-day rental, so he was charged an additional \$2.99 (it was 5 days late, so we could have charged him \$4.98 in late fees, but we’ll keep it simple and just charge the basic fee as the late fee).

We've broken the problem into separate parts so they can be implemented and tested separately.  Your job is to fill in the code cells so each part passes the auto-grader tests.

**Note:** &nbsp; There is no documentation requirement for this project.

### <span style="color:teal">Database Connection</span>

To simplify the code for this program you can define the database connection as a global variable.  The following code cell imports the libraries you need and defines the database connection.

In [23]:
from sqlite3 import connect

db = connect('sakila211.db')

### <span style="color:teal">Dates</span>

Dates in a SQLite database are stored as strings.  The function below uses a Python library named `datetime` to parse a date string and return a string that can be printed as part of the report.  Read the online documentation for the `datetime` module if you want to learn more about date formats and functions that parse and format dates.

**Note:** &nbsp; The format string is a global variable -- you can use this same format in your function that checks to see if a CD is overdue.

In [24]:
from datetime import datetime

date_format = "%Y-%m-%d %H:%M:%S.%f"

def mmddyyyy(date):
    d = datetime.strptime(date, date_format)
    return '{:02d}/{:02d}/{:02d}'.format(d.month, d.day, d.year)

In [25]:
assert mmddyyyy('2005-02-26 20:59:35.000') == '02/26/2005'

### <span style="color:teal">Customer Name</span>

Fill in the definition of the `customer_name` function.  It will be passed a string with a last name, and it should return a list of tuples of the form `(first, last)` with the first and last names of all customers with the same last name.  If there are no customers with that name the list should be empty.

In [35]:
name = db.execute("SELECT first_name, last_name FROM customer LIMIT 5")
print(tup[0] from tup in name)

SyntaxError: invalid syntax (<ipython-input-35-d009a153cf4f>, line 2)

In [30]:
def customer_names(s):
    names = db.execute("SELECT first_name, last_name FROM customer WHERE last_name = ? ")
    return names

customer_names('ebert')

ProgrammingError: Incorrect number of bindings supplied. The current statement uses 1, and there are 0 supplied.

In [28]:
def customer_names(s):
    names = db.execute("SELECT first_name, last_name FROM customer WHERE last_name = ? ")
    for row in names:
        first, last = row
        if last.upper == s.upper:
            return list(tuple(row))

##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [29]:
assert customer_names('ebert') == [('Leo', 'Ebert')]

ProgrammingError: Incorrect number of bindings supplied. The current statement uses 1, and there are 0 supplied.

In [73]:
assert customer_names('smith') == [('Mary', 'Smith')]

AssertionError: 

In [74]:
assert customer_names('turing') == []

AssertionError: 

### <span style="color:teal">Film List</span>

Fill in the body of the `film_list` function so it returns a list of all CDs rented by a customer during a specified month. 
The parameters that will be passed to the function are the customer's last name (a string) and the month and year (both integers).  Return a list of tuples that have the following values, in the order shown:
* film title
* rental rate (how much it costs to rent the CD)
* rental date (the date the CD was rented)
* return date (the date the CD was returned)
* rental duration (the number of days the CD can be out before a late fee is charged) 

All of the information you need can be retrieved with a single query that joins the rental, inventory, film, and customer tables. Note that this is the same query used to answer Question 19 above, except with different column names.

**Hint:** &nbsp; The dates in the database are strings.  Think about pattern matching when you write the WHERE clause that specifies the date range.


In [93]:
! sqlite3 sakila211.db "SELECT title, rental_rate, rental_date, return_date, rental_duration FROM film JOIN inventory USING (film_id) JOIN rental USING (inventory_id)"

title            rental_rate  rental_date              return_date              rental_duration
---------------  -----------  -----------------------  -----------------------  ---------------
Blanket Beverly  2.99         2005-05-24 22:53:30.000  2005-05-26 22:04:30.000  7              
Freaky Pocus     2.99         2005-05-24 22:54:33.000  2005-05-28 19:40:33.000  7              
Graduate Lord    2.99         2005-05-24 23:03:39.000  2005-06-01 22:12:39.000  7              
Love Suicides    0.99         2005-05-24 23:04:41.000  2005-06-03 01:43:41.000  6              
Idols Snatchers  2.99         2005-05-24 23:05:21.000  2005-06-02 04:33:21.000  5              
Mystic Truman    0.99         2005-05-24 23:08:07.000  2005-05-27 01:32:07.000  5              
Swarm Gold       0.99         2005-05-24 23:11:53.000  2005-05-29 20:34:53.000  4              
Lawless Vision   4.99         2005-05-24 23:31:46.000  2005-05-27 23:33:46.000  6              
Matrix Snowman   4.99         2005-05-25

In [114]:
def film_list(name, month, year):
    query = "SELECT title, rental_rate, rental_date, return_date, rental_duration FROM film JOIN inventory USING (film_id) JOIN rental USING (inventory_id) WHERE rental_date[1] == '{}'"
    result = db.execute(query, (month))
    for row in result:
        title, rate, rental_date, return_date, duration = row
        return row

##### <span style="color:red">Autograder Tests:</span>

**Do not edit, copy, or delete these cells.**

In [115]:
leo = film_list('Ebert', 6, 2005)
assert len(leo) == 5
assert leo[0] == (('Lonely Elephant', 2.99, '2005-06-16 20:59:35.000', '2005-06-24 15:31:35.000', 3))

OperationalError: near "[1]": syntax error

In [117]:
mary = film_list('Smith', 8, 2005)
assert len(mary) == 11
assert mary[0] == ('Adaptation Holes', 2.99, '2005-08-01 08:51:04.000', '2005-08-10 12:12:04.000', 7)

OperationalError: near "[1]": syntax error

### <span style="color:teal">Overdue CDs</span>

To decide whether a CD was returned late you need to subtract the rental date from the return date and see if the difference is greater than the rental duration. The easiest way to do this is to use Python’s `datetime` library.

Using a call to `strptime` (see the body of the `mmddyyyy` function defined above) create `datetime` objects for the rental date and the return date. Then subtract the rental date from the return date to get a `timedelta` object:
```
    diff = returned - rented
```
Now you can get attributes of the object named `diff` to tell you how far apart the dates were, for example `diff.days` will tell you the time difference in days.

Fill in the body of the `overdue` function.  The parameters are a rental date, a return date, and the rental duration.  Return TRUE if the CD was not returned or if the number of days it was checked out is greater than the rental duration.

In [ ]:
def overdue(rent_date, return_date, duration):
    # YOUR CODE HERE
    pass

In [ ]:
assert overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:35.000', 2)

In [ ]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-17 20:59:35.000', 2)

In [ ]:
assert not overdue('2005-06-16 20:59:35.000', '2005-06-19 20:59:34.000', 2)

### <span style="color:teal">Report Generator</span>

Fill in the body of the function named `sakila_report` so it prints a monthly report.  The arguments passed to the function will be a customer's last name (a string) and a month and year (both integers).

If the customer is not in the database, or if there are two customers with the same last name, return None.  Otherwise print a list of films and late fees and the total amount due.

You can format the report however you would like, as long as the following items appear somewhere in the output:
* every fee should be printed with a dollar sign and an amount, for example `$2.99` or `25.92`
* the total amound due should be printed on a line that includes the word `"Total"` followed by a dollar amount on the same line

In [ ]:
def sakila_report(customer, month, year):
    # YOUR CODE HERE
    pass

##### <span style="color:red">Tests</span>

You can use code cells in this section (and add more if you wish) to test your functions.

##### <span style="color:red">Autograder Tests</span>

The tests for the `sakila_report` function will call the function and while using the IPython `capture` command to save everything the function prints.  We'll then use a function that counts the number of times a dollar amount (from either a rental fee or a late fee) occurs and make sure the total amount is the expected value.

**Do not edit, copy, or delete these cells.**

In [ ]:
import re

def check_report(text, expected_count, expected_total):
    nums = []
    for line in text.split('\n'):
        has_total = re.search(r'Total.*(\$\d+\.\d\d)', line)
        if has_total:
            assert has_total.group(1) == expected_total
        else:
            for s in re.findall(r'\$\d+\.\d\d', line):
                nums.append(s)
    assert(len(nums) == expected_count)
    total = 0.0
    for s in nums:
        total += float(s[1:])
    assert(round(total,2)) == round(float(expected_total[1:]), 2)
    return True

In [ ]:
%%capture res
sakila_report('Ebert', 6, 2005)

In [ ]:
assert check_report(res.stdout, 8, '$25.92')

In [ ]:
%%capture res
sakila_report('Smith', 8, 2005)

In [ ]:
assert check_report(res.stdout, 15, '$32.85')

In [ ]:
%%capture res
sakila_report('Turing', 6, 2005)

In [ ]:
assert len(res.stdout) == 0

In [ ]:
sakila_report('Ebert', 6, 2005)